src/coreclr/vm/jithelpers.cpp

```diff
@@ -533,77 +533,17 @@ HCIMPL1(void*, JIT_GetStaticFieldAddr, FieldDesc* pFD)
 HCIMPLEND
 #include <optdefault.h>
 
-// Slow helper to tailcall from the fast one
-NOINLINE HCIMPL1(void, JIT_InitClass_Framed, MethodTable* pMT)
-{
-    FCALL_CONTRACT;
-
-    HELPER_METHOD_FRAME_BEGIN_0();
-
-    // We don't want to be calling JIT_InitClass at all for perf reasons
-    // on the Global Class <Module> as the Class loading logic ensures that we
-    // already have initialized the Global Class <Module>
-    CONSISTENCY_CHECK(!pMT->IsGlobalClass());
-
-    _ASSERTE(pMT->IsFullyLoaded());
-    pMT->CheckRunClassInitThrowing();
-
-    HELPER_METHOD_FRAME_END();
-}
-HCIMPLEND
-
-
-/*************************************************************/
-#include <optsmallperfcritical.h>
-HCIMPL1(void, JIT_InitClass, CORINFO_CLASS_HANDLE typeHnd_)
-{
-    FCALL_CONTRACT;
-
-    TypeHandle typeHnd(typeHnd_);
-    MethodTable *pMT = typeHnd.AsMethodTable();
-
-    if (pMT->IsClassInited())
-        return;
-
-    // Tailcall to the slow helper
-    ENDFORBIDGC();
-    HCCALL1(JIT_InitClass_Framed, pMT);
-}
-HCIMPLEND
-#include <optdefault.h>
-
-/*************************************************************/
-HCIMPL2(void, JIT_InitInstantiatedClass, CORINFO_CLASS_HANDLE typeHnd_, CORINFO_METHOD_HANDLE methHnd_)
+// Helper for the managed InitClass implementations
+extern "C" void QCALLTYPE InitClassHelper(MethodTable* pMT)
 {
-    CONTRACTL {
-        FCALL_CHECK;
-        PRECONDITION(methHnd_ != NULL);
-    } CONTRACTL_END;
-
-    HELPER_METHOD_FRAME_BEGIN_NOPOLL();    // Set up a frame
-
-    MethodTable * pMT = (MethodTable*) typeHnd_;
-    MethodDesc *  pMD = (MethodDesc*)  methHnd_;
-
-    MethodTable * pTemplateMT = pMD->GetMethodTable();
-    if (pTemplateMT->IsSharedByGenericInstantiations())
-    {
-        pMT = ClassLoader::LoadGenericInstantiationThrowing(pTemplateMT->GetModule(),
-                                                            pTemplateMT->GetCl(),
-                                                            pMD->GetExactClassInstantiation(pMT)).AsMethodTable();
-    }
-    else
-    {
-        pMT = pTemplateMT;
-    }
+    QCALL_CONTRACT;
+    BEGIN_QCALL;
 
     _ASSERTE(pMT->IsFullyLoaded());
     pMT->EnsureInstanceActive();
```

Comment:
Why did we need to call `EnsureInstanceActive` here and not in the regular `JIT_InitClass`?

I do not see why `EnsureInstanceActive` should be necessary here. It looks like a workaround for a bug somewhere else in the system.

Reply:
I'm pretty sure the entire approach to how activation works is not entirely reliable. Notably, since logic running in the Module constructor is allowed to permit a call to `EnsureInstanceActive` to complete without the module constructor finishing, in theory we need to be EXTREMELY careful in the runtime data structures to avoid allowing any escapes of objects produced from the module constructor from becoming embedded in the state of objects which are not constrained to be accessed only within that module constructor. However, we don't have such protections. For instance, if during the module constructor execution, an module defined object `ModuleDefined` with a method `M` is both called in the module constructor, and converted to a delegate, and the delegate stashed in an object reachable from the static variables of some OTHER module, then we can execute code from a not yet active module by calling that delegate.

In practice, this doesn't appear to be a major problem, and our rules are currently strong enough to protect the actual execution behavior of module constructors, and the static field members of a module, but I don't think the current model is actually strictly "correct".

In this particular case, I expect that the old `JIT_InitInstantiatedClass` call to `EnsureInstanceActive` was making the system a bit more robust around reliably activating the module, but not completely so.

Overall, this PR strengthens the behavior of `EnsureInstanceActive` around generics (and non-generics) in a few ways, which I'm not convinced is entirely necessary, but I don't think will break things. These changes, are driven by the desire to reduce duplicate code paths (this change, which makes the old JIT_InitClass ALSO force the instance to be active), and makes generics parameters of base types active when instantiating a derived type as well as making generic interfaces have similar properties.